### DISCUSSION TITLE
Is it possible to retrieve InterfaceDeclarationSyntax from TypeSyntax (e.g. ReturnType for MethodDeclarationSyntax)?

### DISCUSSION THREAD

**Discussion by pedro-gilmora** [dotnet/roslyn#64933] [Q&A]
Hi, I'm just trying to create a source generator

I want to know if it's posible to get an interface declaration being a methods's return type.

My idea is implement a default body for an interface method, that one should be a generated class which implements the method's return type, that should be an interface

```cs
private string BuildMethod(MethodDeclarationSyntax method)
{
  string 
    typeName = method.ReturnType.ToString(), //Should return an interface type
    identifier = method.Identifier.ValueText,
    paramsDef = string.Join(", ", method
        .ParameterList
        .Parameters
        .Select(ip => $"{ip.Type} {ip.Identifier}{ip.Default}" ) ),
    paramsRef = string.Join(", ", method
        .ParameterList
        .Parameters
        .Select(ip => ip.Identifier.ValueText ) );
  
  this.typesToBeGenerated.Add( GetInterfaceDeclarationSyntax( method.ReturnType ) ); // <= This is the key of the question 
  
  //typeName[1..] Should be the type to generate based on its interface
  return $"public { typeName } { method.Identifier }{ method.TypeParameterList }({ paramsDef }) => new { typeName[1..] }(this, { paramsRef });";
}

private static string GetInterfaceDeclarationSyntax(TypeSyntax _interface){
   // Logic to retrieve interface definition
}
```

I guess the idea it's not wrong at all. Any help will be appreciated in advance.

**Comment by CyrusNajmabadi** [comment]
> GetInterfaceDeclarationSyntax( method.ReturnType ) ); // <= This is the key of the question 

@pedro-gilmora you'll have to provide more information as to what isn't working here.  For determining if a Source Generators (SGs) are an applicable solution to a problem, just ask the question:

1. could the user solve this problem by adding a new source file with C# code it in to the rest of their code and compiling it all together?

If the answer is 'yes' then you can use SGs to solve it as well.  An SG just spits out source files to compile.  So if you know what you would want in those source files, then you can make the SG spit that out.

**Reply by pedro-gilmora** [reply]
Let's say I have some interface like this one. I'll change methods return type for properties and indexers types.

>could the user solve this problem by adding a new source file with C# code it in to the rest of their code and compiling it all together?

The idea is to create a default implementation for `IStoreSdk`, `ICustomerHandler` and `ICustomerActionsById` so the developer doesn't have to create one

```cs
public interface IStoreSdk {

  // ⬇️Interface definition to be found by the Roslyn Analyzer
  ICustomerHandler Customers { get; }

}

public interface ICustomerHandler {

  Task<Customer> CreateAsync(Customer initCustomer);

  // ⬇️Interface definition to be found by the Roslyn Analyzer
  ICustomerActionsById this[long id] { get; }

}

  // ⬇️This is the interface declaration
public interface ICustomerActionsById {

  Task<Customer> GetAsync();

  Task<ISdkResponse> UpdateAsync(Customer customer);

  Task<ISdkResponse> DeleteAsync(Customer customer);

} 
```

It should generate the following classes under some specs and conventions (of course, in separate files):

```cs
public sealed StoreSdk {

  CustomerHandler Customers => new CustomerHandler(this)

}

public sealed class CustomerHandler: ICustomerHandler {

  private readonly StoreSdk _storeSdk;

  public CustomerHandler(StoreSdk  storeSdk) 
     => this._storeSdk;

  public Task CreateAsync(Customer initCustomer) 
     => this._storeSdk.Create(initCustomer);

  public ICustomerActionsById this[long id]  
     => new CustomerActionsById(this._storeSdk, id);

}

public class CustomerActionsById: ICustomerActionsById {

  private readonly StoreSdk _storeSdk;
  private readonly long _id;


  public CustomerActionsById(StoreSdk  storeSdk, long id)
     => (_storeSdk_id) = (storeSdk, id);

  public Task<Customer> GetAsync() 
     =>  _storeSdk.GetAsync<Customer>(new {id: _id});

  public Task<ISdkResponse> UpdateAsync(Customer customer) 
     => _storeSdk.UpdateAsync<ISdkResponse>(new {id: _id}, initCustomer);

  public Task<ISdkResponse> DeleteAsync(Customer customer) 
     => _storeSdk.DeleteAsync<ISdkResponse>(new {id: _id}, initCustomer);

} 
```

I hope you get it

**Reply by CyrusNajmabadi** [reply]
So if the user could do this by adding all that code themselves into a new file, then it's something you can do in a source generator.

**Reply by pedro-gilmora** [reply]
> So if the user could do this by adding all that code themselves into a new file, then it's something you can do in a source generator.

The thing is, how to retrieve the declaration of all those nested-interface-references ?

**Reply by CyrusNajmabadi** [reply]
Looks like how you're doing it here:

> this.typesToBeGenerated.Add( GetInterfaceDeclarationSyntax( method.ReturnType ) ); // <= This is the key of the question 

Why is this not sufficient?

**Reply by CyrusNajmabadi** [reply]
note: for general q/a, i'd recommend going to gitter, like discord.gg/csharp.  There are channels there for things like Roslyn where you can ask source-generator questions :)

**Reply by pedro-gilmora** [reply]
> Looks like how you're doing it here:
> 
> > this.typesToBeGenerated.Add( GetInterfaceDeclarationSyntax( method.ReturnType ) ); // <= This is the key of the question
> 
> Why is this not sufficient?

Because what I need is the code to retrieve the interface definition/declaration from that method return type

**Reply by CyrusNajmabadi** [reply]
There is no interface declaration. You'd have to generate it.
